### PR TITLE
Multi episode value column calculation

### PR DESCRIPTION
**Pull request by Laurenstc**
*Issue #, if available:*

*Description of changes:*
- Allow automatic calculation of q-value for WIDataframe for multi process problems

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/a2rl/_dataframe.py [hunk 1]
```diff
@@ -679,6 +679,110 @@ def add_value(
...
+def add_value_for_multi_episode_process(
+    self,
+    sarsa: bool = True,
+    alpha: float = 0.1,
+    gamma: float = 0.6,
+    value_col: str = "value",
+    override: Literal["replace", "warn", "error"] = "replace",
+    episode_identifier: str = "",
+) -> WiDataFrame:
+    """Append column ``value_col`` into this dataframe (restriction: ``df`` must NOT contain
```
Comment (verdimrc) [inline src/a2rl/_dataframe.py:691]:
```suggestion
    """Return a new dataframe with additional column ``value_col`` (restriction: ``df`` must NOT contain
```

File: src/a2rl/_dataframe.py [hunk 2]
```diff
@@ -679,6 +679,110 @@ def add_value(
...
+        gamma: Discount factor of future reward in Q-Learning and SARSA. Must be within 0 and 1.
+        value_col: The column name for the computed values.
+        override: What to do when this dataframe has had column ``value_col``. Valid values
+            are ``replace`` to silently override, ``warn`` to show a warning, and ``raise`` to
+            raise a :exc:`ValueError`.
+        episode_identifier: identifier in the 'self' dataframe that allows for groupby and
+            identifies a process.
+
+    Returns:
+        This dataframe, modified with an additional ``value_col`` column. This return value is
```
Comment (verdimrc) [inline src/a2rl/_dataframe.py:712]:
```suggestion
        A new dataframe with an additional ``value_col`` column. This return value is
```

File: src/a2rl/_dataframe.py [hunk 3]
```diff
@@ -679,6 +679,110 @@ def add_value(
...
+            q_table[df_t["_state"].astype(int), df_t["_action"].astype(int)],
+            index=df.index,
+        )
+
+    x = df.groupby(episode_identifier).apply(lambda x: calc_q_values(x, df, df_t, alpha, gamma))
+    x2 = x.reset_index(level=0)
+    x2.rename({0: "value"}, axis=1, inplace=True)
+    x2.drop([episode_identifier], axis=1, inplace=True)
+    merged = pd.merge(df, x2, how="left", left_index=True, right_index=True)
+    return merged
```
Comment (verdimrc) [inline src/a2rl/_dataframe.py:783]: Propose to change the behavior to modify in-place -- the same semantic in the single-episode counterpart.

```suggestion
    # len_before_modif = self.shape[0]
    self[value_col] = merged[value_col]  # Make sure index matches
    # assert self.shape[0] == len_before_modif
    return self
```

However, if this breaks too much, I'm happy to accept this version first once the docstring (description & returns) is updated to say that a new dataframe is returned.

**Review by verdimrc — CHANGES_REQUESTED** [review]
Thanks for this PR to address a frequent ask. Could you also add a unit test for this new function?

**Comment by Laurenstc** [comment]
> add a unit test 
I've attempted to add a test - not entirely sure whether this is as expected! Let me know @verdimrc

**Comment by Laurenstc** [comment]
> Thanks for this PR to address a frequent ask. Could you also add a unit test for this new function?

I've added the tests! Unsure whether my new fixtures are running properly thought!

File: src/a2rl/_dataframe.py [hunk 4]
```diff
@@ -677,6 +677,112 @@ def add_value(
             self._rewards[1] = value_col
         return self
 
+    def add_value_for_multi_episode_process(
+        self,
+        sarsa: bool = True,
+        alpha: float = 0.1,
+        gamma: float = 0.6,
+        value_col: str = "value",
+        episode_identifier: str = "episode",
```
Comment (verdimrc) [inline src/a2rl/_dataframe.py:686]: Changed default to non-empty string.

File: src/a2rl/_dataframe.py [hunk 5]
```diff
@@ -677,6 +677,112 @@ def add_value(
...
+                for i in range(df_t.index[0], df_t.index[-1]):
+                    state = int(df_t.loc[i, "_state"])
+                    next_state = int(df_t.loc[i + 1, "_state"])
+                    action = int(df_t.loc[i, "_action"])
+                    reward = df_t.loc[i + 1, "_reward"]
+                    old_value = q_table[state, action]
+
+                    if sarsa:
+                        next_value = q_table[next_state, np.argmax(q_table[next_state])]
+                        new_value = old_value + alpha * (reward + gamma * next_value - old_value)
```
Comment (verdimrc) [inline src/a2rl/_dataframe.py:760]: Remove the (1-alpha) term from `old_value`. Aligned with canonical q-value or sarsa formula, and in-line with PR #21 which fixes `add_value()`.

**Review by verdimrc — APPROVED** [review]
@Laurenstc: your changes approved.

I also proposed some changes, kindly review if they make sense (and revert if not). Whatever the case, you're good to hit the merge button :)

**Pull request merged**